src/relax/transform/infer_layout_utils.h

```diff
@@ -0,0 +1,201 @@
+ * \brief Utility functions to alter the layouts of operators or replace primitive operators with
+          other expressions. This pass can be used for computing convolution in
+          custom layouts or other general weight pre-transformation.
+ */
+
+#ifndef TVM_RELAX_TRANSFORM_INFER_LAYOUT_UTILS_H_
+#define TVM_RELAX_TRANSFORM_INFER_LAYOUT_UTILS_H_
+
+#include <tvm/relax/attrs/create.h>
+#include <tvm/relax/attrs/datatype.h>
+#include <tvm/relax/attrs/image.h>
+#include <tvm/relax/attrs/linear_algebra.h>
+#include <tvm/relax/attrs/memory.h>
+#include <tvm/relax/attrs/nn.h>
+#include <tvm/relax/attrs/shape.h>
+#include <tvm/relax/attrs/statistical.h>
+#include <tvm/relax/expr.h>
+#include <tvm/relax/nested_msg.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/tir/data_layout.h>
+
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "utils.h"
+
+namespace tvm {
+namespace relax {
+
+using tir::Layout;
+using NLayout = NestedMsg<Layout>;
+
+/*
+ * \brief An output structure to hold results from FInferCorrectLayout calls.
+ * \tparam input_layouts Inferred input layouts.
+ * \tparam output_layouts Inferred output layouts.
+ * \tparam new_attrs Updated attributes consistent with inferred layouts.
+ */
+class InferLayoutOutputNode : public Object {
+ public:
+  Array<NLayout> input_layouts;
+  Array<NLayout> output_layouts;
+  Attrs new_attrs;
+
+  void VisitAttrs(tvm::AttrVisitor* v) {
+    v->Visit("input_layouts", &input_layouts);
+    v->Visit("output_layouts", &output_layouts);
+    v->Visit("new_attrs", &new_attrs);
+  }
+
+  TVM_DECLARE_BASE_OBJECT_INFO(InferLayoutOutputNode, Object);
+
+  static constexpr const char* _type_key = "relax.transform.InferLayoutOutput";
+};
+
+class InferLayoutOutput : public ObjectRef {
+ public:
+  InferLayoutOutput(Array<NLayout> input_layouts, Array<NLayout> output_layouts, Attrs new_attrs) {
+    auto n = make_object<InferLayoutOutputNode>();
+    n->input_layouts = std::move(input_layouts);
+    n->output_layouts = std::move(output_layouts);
+    n->new_attrs = std::move(new_attrs);
+    data_ = n;
+  }
+  TVM_DEFINE_OBJECT_REF_METHODS(InferLayoutOutput, ObjectRef, InferLayoutOutputNode);
+};
+
+struct NLayoutEqual {
+  bool operator()(const NLayout& a, const NLayout& b) const {
+    auto layout_equal = [](const Layout& a, const Layout& b) {
+      if (a.defined() && b.defined()) {
+        return a.name() == b.name();
+      } else {
+        return a.defined() == b.defined();
+      }
+    };
+    return Equal(a, b, layout_equal);
+  }
+};
+
+using VarLayoutMap = Map<Var, NLayout>;
+
+/*!
+ * \brief Layout convertion inferface.
+ * \param call The call node.
+ * \param desired_layouts The desired layouts of the operator.
+ * \param var_layout_map The layout of the variables.
+ */
+using FRelaxInferLayout = runtime::TypedPackedFunc<InferLayoutOutput(
+    Call call, Map<String, Array<String>> desired_layouts, VarLayoutMap var_layout_map)>;
```

Comment:
Do you think it a good idea to mark them with constant reference? Given we won’t do change to these inputs.